### PR TITLE
Feat/NewArticle

### PR DESCRIPTION
**Pull request by joaooliveiradev**

### CODE REVIEW

**Comment by joaooliveiradev** [comment]
- Feat: Some adjusts to better focus management.
- Refactor: Adjust on stories of several components.
- Refactor: Removing toast context
- Feat: Adding new AuthError class to handling errors globally.
- Fix: Adding more options in the destroy cookies function to really removes the auth cookie from the cookies.
- Feat: Adding Global Error Handling, logout the user if token it's not authorized.
- Feat: Adjust Sign In and Sign Up in a single component inside SignInModal.
- Feat: Adjust the options for the useQuery hooks.
- Refactor: Putting the Loading component inside Button Component.
- Feat: Making the box shadow colors default.
- Refactor: adjust some style issues on Layout component.
- Chore: putting some dependencies to exact.
- Fix: Solve cache problem on useMe hook.
- Feat: Putting focus on the first input in both of the forms of SignInModal.
- Fix: More design issues on Article Grid, Article Card, Footer, Header, Layout, Button.
- Fix: Adjust Toast component that doesn't close on enter key.
- Refactor: Removing Global Styles from Storybook.
- Fix: design issues on Hero Component.
- Feat: Adding React Query eslint config to ensure best practices.
- Feat: Adding the a11y and styled-a11y eslint plugin to help develop more accessible components.
- Feat: Putting some outline values back to help focus.
- Refactor: Removing unnecessary folders queries/mutations
- Feat: Adding stories for Hero/Divider components.
- Feat: Adding a function that goes back the focus to header component.
- Feat: Adding a condition to prevent unnecessary fetch when doesn't have another page on infinite scroll.
- Feat: Adding TextArea Component
- Feat: Adding basic implementation of editor page to create new article.
- Feat: Making useNewArticle Mutation hook

**Comment by joaooliveiradev** [comment]
Feat: Some adjusts to better focus management.
Refactor: Adjust on stories of several components.
Refactor: Removing toast context
Feat: Adding new AuthError class to handling errors globally.
Fix: Adding more options in the destroy cookies function to really removes the auth cookie from the cookies.
Feat: Adding Global Error Handling, logout the user if token it's not authorized.
Feat: Adjust Sign In and Sign Up in a single component inside SignInModal.
Feat: Adjust the options for the useQuery hooks.
Refactor: Putting the Loading component inside Button Component.
Feat: Making the box shadow colors default.
Refactor: adjust some style issues on Layout component.
Chore: putting some dependencies to exact.
Fix: Solve cache problem on useMe hook.
Feat: Putting focus on the first input in both of the forms of SignInModal.
Fix: More design issues on Article Grid, Article Card, Footer, Header, Layout, Button.
Fix: Adjust Toast component that doesn't close on enter key.
Refactor: Removing Global Styles from Storybook.
Fix: design issues on Hero Component.
Feat: Adding React Query eslint config to ensure best practices.
Feat: Adding the a11y and styled-a11y eslint plugin to help develop more accessible components.
Feat: Putting some outline values back to help focus.
Refactor: Removing unnecessary folders queries/mutations
Feat: Adding stories for Hero/Divider components.
Feat: Adding a function that goes back the focus to header component.
Feat: Adding a condition to prevent unnecessary fetch when doesn't have another page on infinite scroll.
Feat: Adding TextArea Component
Feat: Adding basic implementation of editor page to create new article.
Feat: Making useNewArticle Mutation hook

Feat: Add Radix Acessibility Icons.
Refactor: Exporting SVGs by components folder.
Feat: Check Icon
Feat: Adding Validationg on Submit in NewArticle form.
Feat: New Alert component for inline.
Feat: Adding feedback inline in NewArticle form with Alert component.
Feat: TextArea Component.
Refactor: Change in some imports and css function from styled components in all the application.
Feat: Adding a logic to persist data on NewArticle.
Refactor: Changing to late validation the both SignIn and SignUp form.
Refactor: Removing touched prop in Input component and TextArea component.
Feat: Adding a Label Screen Reader only in Input and Textarea.
Refactor: Some refactor on handling the error on NewArticle, Header, SignInModal, ArticlePage, ProfilePage
Refactor: Some refactors along of the application using a better composition of functions using fp ts.
Refactor: Implementing the highlight of the code using rehype-highlight to reduce the bundle.
Refactor: Deleting automatic prefetch of the page using Next Links.
Feat: Disabling side effects on some types of files in Webpack config to make them theshakable and reduce the bundle.
Feat: Adding some security headers to mitigate some attacks.
Feat: Adding a basic implementation Cache Control header in some pages.
Chore: Updating Radix Packages to the latest version to receive a better compatibility with esm.
Chore: Update Storybook to version 7.
Refactor: Moving calculateTotalArticles function to inside of useGetArticles.
Feat: Adding stories for SignIn, SignUp
Refactor: Changes on handling the SEO

**Comment by joaooliveiradev** [comment]
Feat: Some adjusts to better focus management.
Refactor: Adjust on stories of several components.
Refactor: Removing toast context
Feat: Adding new AuthError class to handling errors globally.
Fix: Adding more options in the destroy cookies function to really removes the auth cookie from the cookies.
Feat: Adding Global Error Handling, logout the user if token it's not authorized.
Feat: Adjust Sign In and Sign Up in a single component inside SignInModal.
Feat: Adjust the options for the useQuery hooks.
Refactor: Putting the Loading component inside Button Component.
Feat: Making the box shadow colors default.
Refactor: adjust some style issues on Layout component.
Chore: putting some dependencies to exact.
Fix: Solve cache problem on useMe hook.
Feat: Putting focus on the first input in both of the forms of SignInModal.
Fix: More design issues on Article Grid, Article Card, Footer, Header, Layout, Button.
Fix: Adjust Toast component that doesn't close on enter key.
Refactor: Removing Global Styles from Storybook.
Fix: design issues on Hero Component.
Feat: Adding React Query eslint config to ensure best practices.
Feat: Adding the a11y and styled-a11y eslint plugin to help develop more accessible components.
Feat: Putting some outline values back to help focus.
Refactor: Removing unnecessary folders queries/mutations
Feat: Adding stories for Hero/Divider components.
Feat: Adding a function that goes back the focus to header component.
Feat: Adding a condition to prevent unnecessary fetch when doesn't have another page on infinite scroll.
Feat: Adding TextArea Component
Feat: Adding basic implementation of editor page to create new article.
Feat: Making useNewArticle Mutation hook

Feat: Add Radix Acessibility Icons.
Refactor: Exporting SVGs by components folder.
Feat: Check Icon
Feat: Adding Validationg on Submit in NewArticle form.
Feat: New Alert component for inline.
Feat: Adding feedback inline in NewArticle form with Alert component.
Feat: TextArea Component.
Refactor: Change in some imports and css function from styled components in all the application.
Feat: Adding a logic to persist data on NewArticle.
Refactor: Changing to late validation the both SignIn and SignUp form.
Refactor: Removing touched prop in Input component and TextArea component.
Feat: Adding a Label Screen Reader only in Input and Textarea.
Refactor: Some refactor on handling the error on NewArticle, Header, SignInModal, ArticlePage, ProfilePage
Refactor: Some refactors along of the application using a better composition of functions using fp ts.
Refactor: Implementing the highlight of the code using rehype-highlight to reduce the bundle.
Refactor: Deleting automatic prefetch of the page using Next Links.
Feat: Disabling side effects on some types of files in Webpack config to make them theshakable and reduce the bundle.
Feat: Adding some security headers to mitigate some attacks.
Feat: Adding a basic implementation Cache Control header in some pages.
Chore: Updating Radix Packages to the latest version to receive a better compatibility with esm.
Chore: Update Storybook to version 7.
Refactor: Moving calculateTotalArticles function to inside of useGetArticles.
Feat: Adding stories for SignIn, SignUp
Refactor: Changes on handling the SEO

Refactor: improvements in Hero Component (getting the user state by hook)
Feat: Add 404 page and unauthorized page.
Feat: Validation to display the edit button and new article button only in the current user profile page, not the public ones.

File: src/components/Alert/Alert.tsx [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        </AlertContext.Provider>
+      </Wrapper>
+    )
+  } else return null
+}
+
+const Icon = () => {
+  const { status } = useContext(AlertContext)
+  return status === 'error' ? <ExclamationIcon /> : <CheckIcon />
+}
```
Comment (vmarcosp) [inline src/components/Alert/Alert.tsx:121]: Tratar o status com switch case ao inves de if.

File: src/components/Alert/Alert.tsx [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+Alert.Icon = Icon
+
+export type TextProps = {
```
Comment (vmarcosp) [inline src/components/Alert/Alert.tsx:125]: Alterar para PropsWithChildren

File: src/components/ArticleBody/ArticleBody.tsx [hunk 3]
```diff
@@ -75,37 +66,32 @@ const Wrapper = styled.section`
...
-    >
-      {String(children)}
-    </CodeHighlighter>
-  )
+const rehypeSanitizeSettings: RehypeSanitizeOptions = {
+  ...defaultSchema,
+  attributes: {
+    ...defaultSchema.attributes,
+    code: [
+      ...((defaultSchema.attributes && defaultSchema.attributes.code) || []),
```
Comment (vmarcosp) [inline src/components/ArticleBody/ArticleBody.tsx:81]: Separar em uma const/var separada e com um nome declarativo.

File: src/components/ArticleStats/ArticleStats.tsx [hunk 4]
```diff
@@ -40,8 +35,9 @@ export const ArticleStats = ({ readTime, date }: ArticleStatsProps) => {
   }
 
   const getDate = (articleDate: string) => {
-    const formatMethod = 'PP'
-    const date = format(new Date(articleDate), formatMethod)
+    dayjs.extend(utc)
```
Comment (vmarcosp) [inline src/components/ArticleStats/ArticleStats.tsx:38]: Melhorar composicao de componentes (ArticleStats, ArticleDate, ArticleReadime, etc),

File: src/components/Dropdown/Dropdown.tsx [hunk 5]
```diff
@@ -88,22 +90,23 @@ const DropdownButton = styled.button`
...
-    :focus-visible {
-      box-shadow: 0 0 0 2px ${transparentize(0.6, theme.colors.black[200])};
+    &:focus-visible {
+      outline: 1px solid transparent;
+      box-shadow: 0 0 0 2px ${theme.colors.black[200]};
     }
   `}
 `
 
 export const Dropdown = ({ children, trigger }: DropdownProps) => {
```
Comment (vmarcosp) [inline src/components/Dropdown/Dropdown.tsx:100]: Sugestao de melhoria:
```typescript
 <Dropdown>
          <DropdownTrigger>
            <ProfileName size={2} name={maybeData.value.user.username} />
            </DropdownTrigger>
          <DropdownContent>
            <DropdownList>
            <DropdownItem
            href={`/profile/${maybeData.value.user.username}`}
            label="Profile"
          />
          <DropdownItem label="Sign Out" onEventClick={signOut} />
            </DropdownList>
          </DropdownContent>
        </Dropdown>
```

File: src/components/EmptyState/EmptyState.tsx [hunk 6]
```diff
@@ -1,41 +1,33 @@
...
-    font-size: ${theme.fonts.sizes.xmedium};
-    color: ${theme.colors.grey[300]};
-    font-weight: 500;
-  `}
+  font-size: ${({ theme }) => theme.fonts.sizes.xmedium};
+  color: ${({ theme }) => theme.colors.grey[300]};
+  font-weight: 500;
 `
 
 export const EmptyState = ({ message, title }: EmptyStateProps) => {
```
Comment (vmarcosp) [inline src/components/EmptyState/EmptyState.tsx:33]: Melhorar composicao.

File: src/components/ErrorState/ErrorState.tsx [hunk 7]
```diff
@@ -61,7 +52,7 @@ export const ErrorState = ({
 }: ErrorStateProps) => {
   return (
     <Wrapper>
```
Comment (vmarcosp) [inline src/components/ErrorState/ErrorState.tsx:54]: Melhorar composicao.

File: src/components/Footer/Footer.tsx [hunk 8]
```diff
@@ -32,21 +29,37 @@ const Text = styled.p`
...
-  ${({ theme }) => css`
-    color: ${theme.colors.black[100]};
-    border-bottom: 2px solid ${theme.colors.black[100]};
-  `}
+const TextButton = styled(TextButtonDefault)`
+  font-size: ${({ theme }) => theme.fonts.sizes.xmedium};
+  line-height: 23.44px;
 `
 
+const resetFocusToBody = () => {
```
Comment (vmarcosp) [inline src/components/Footer/Footer.tsx:37]: Remover e manter o padrao do browser.

File: src/components/Input/Input.tsx [hunk 9]
```diff
@@ -1,52 +1,88 @@
...
+
+const ErrorFieldMessage = dynamic<ErrorFieldMessageProps>(
+  () =>
+    import('@/components/ErrorFieldMessage/ErrorFieldMessage').then(
+      (module) => module.ErrorFieldMessage
+    ),
+  {
+    ssr: false,
+  }
+)
```
Comment (vmarcosp) [inline src/components/Input/Input.tsx:16]: Testar usando type:
```
export type InputProps = React.InputHTMLAttributes<HTMLInputElement> & {
  errorMessage?: string
  inputRef?: (node: HTMLInputElement | null) => void
}
```

File: src/components/Modal/Modal.tsx [hunk 10]
```diff
@@ -87,19 +90,17 @@ export const Modal = ({
   onOpenChange,
 }: ModalProps) => {
   return (
-    <Dialog.Root open={open} onOpenChange={onOpenChange}>
-      {trigger && <Dialog.Trigger asChild>{trigger}</Dialog.Trigger>}
-      <Dialog.Portal>
+    <Root open={open} onOpenChange={onOpenChange}>
```
Comment (vmarcosp) [inline src/components/Modal/Modal.tsx:93]: Melhorar composicao.

File: src/components/NewArticle/NewArticle.tsx [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+
+const storageKey = 'new-article-editor'
+
+const getStorage = () => {
+  if (typeof window === 'undefined') return initialFieldValues
+
+  const cachedStorage = fromNullable(localStorage.getItem(storageKey))
+
+  return isSome(cachedStorage)
+    ? superJsonParse<NewArticleFieldValues>(cachedStorage.value)
```
Comment (vmarcosp) [inline src/components/NewArticle/NewArticle.tsx:133]: Validar casos de erro e validar o retorno com io-ts.

File: src/components/NewArticle/NewArticle.tsx [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+    : initialFieldValues
+}
+
+const useLocalStorage = () => {
+  const [storage, setStorage] = useState<NewArticleFieldValues>(getStorage)
+
+  useEffect(
+    () => localStorage.setItem(storageKey, superJsonStringify(storage)),
```
Comment (vmarcosp) [inline src/components/NewArticle/NewArticle.tsx:141]: Sugestao: localForage.

File: src/components/NewArticle/NewArticle.tsx [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+    onSubmit: (values) => handleSubmit(values),
+  })
+
+  useEffect(() => {
+    const updatedFieldValues: NewArticleFieldValues = {
+      title: formik.values.title,
+      description: formik.values.description,
+      body: formik.values.body,
+    }
+    setStorage(updatedFieldValues)
```
Comment (vmarcosp) [inline src/components/NewArticle/NewArticle.tsx:176]: setStorage(formik.values)

File: src/components/NewArticle/NewArticle.tsx [hunk 14]
```diff
@@ -0,0 +1,259 @@
...
+  }, [formik.values, setStorage])
```
Comment (vmarcosp) [inline src/components/NewArticle/NewArticle.tsx:177]: Melhroar deps do useEffect.

File: src/components/NewArticle/NewArticle.tsx [hunk 15]
```diff
@@ -0,0 +1,259 @@
...
+
+  return { storage, setStorage }
+}
+
+const firstInputRef = (node: HTMLInputElement | null) =>
+  node ? node.focus() : null
+
+export const NewArticle = () => {
+  const { mutate: createArticle, data, error, isLoading } = useNewArticle()
+  const { storage, setStorage } = useLocalStorage()
```
Comment (vmarcosp) [inline src/components/NewArticle/NewArticle.tsx:153]: Limpar os campos do form pos sucesso na criacao do artigo.

**Review by vmarcosp — CHANGES_REQUESTED** [review]

File: src/pages/editor/index.tsx [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+
+  useEffect(() => {
+    if (!(status === 'loggedIn')) router.replace('/unauthorized')
+  }, [status, router])
+
+  return (
+    <>
+      <NextSeo {...defaultSEO} title="Conduit - New Article" />
+      <Wrapper>
+        <NewArticle />
```
Comment (vmarcosp) [inline src/pages/editor/index.tsx:30]: Mover o NewArticle para ser uma page.

File: src/pages/article/[slug].tsx [hunk 17]
```diff
@@ -70,25 +55,46 @@ const Description = styled.h3`
...
 
-  const maybeArticle = f(() => {
-    const dataOption = fromNullable(data)
-    if (isSome(dataOption) && isRight(dataOption.value)) {
-      return fromEither(dataOption.value)
-    } else return none
-  })
+  const maybeArticle = pipe(data, fromNullable, chain(getRight))
 
   return f(() => {
```
Comment (vmarcosp) [inline src/pages/article/[slug].tsx:60]: Remover f(...)

File: src/components/Alert/Alert.tsx (hunk 1 above)
Comment (joaooliveiradev) [reply src/components/Alert/Alert.tsx]: Fixed in this [commit](https://github.com/joaooliveiradev/conduit-frontend/pull/43/commits/a7f8a75e88fb90247cd3c1a38ee39864f9591d49).

File: src/components/ArticleBody/ArticleBody.tsx (hunk 3 above)
Comment (joaooliveiradev) [reply src/components/ArticleBody/ArticleBody.tsx]: What do you think about [this](https://github.com/joaooliveiradev/conduit-frontend/pull/43/commits/7fa55330498ce6c5f76f83be4bf6582beabe7158)?

File: src/components/ArticleStats/ArticleStats.tsx (hunk 4 above)
Comment (joaooliveiradev) [reply src/components/ArticleStats/ArticleStats.tsx]: Check this [commit](https://github.com/joaooliveiradev/conduit-frontend/pull/43/commits/b4a84412f7e418e76124cc2f7e7a5938c4c42a77).

File: src/components/Dropdown/Dropdown.tsx (hunk 5 above)
Comment (joaooliveiradev) [reply src/components/Dropdown/Dropdown.tsx]: What do you think about this [way](https://github.com/joaooliveiradev/conduit-frontend/pull/43/commits/fb2da6a52a2b63069bfafd44df306a5995fcad8d)?

File: src/components/EmptyState/EmptyState.tsx (hunk 6 above)
Comment (joaooliveiradev) [reply src/components/EmptyState/EmptyState.tsx]: Fixed in this [commit](https://github.com/joaooliveiradev/conduit-frontend/pull/43/commits/dca1f2b4737b9985172e564ce8757510e831e97d).

File: src/components/ErrorState/ErrorState.tsx (hunk 7 above)
Comment (joaooliveiradev) [reply src/components/ErrorState/ErrorState.tsx]: Fixed in this [commit](https://github.com/joaooliveiradev/conduit-frontend/pull/43/commits/3ddd24dab90e99a2fcb1f3a55ce217254f878505)

File: src/components/Footer/Footer.tsx (hunk 8 above)
Comment (joaooliveiradev) [reply src/components/Footer/Footer.tsx]: Fixed in this [commit](https://github.com/joaooliveiradev/conduit-frontend/pull/43/commits/a4766ee9afc626bc1612c6c2796c84051a7e23f2)

File: src/components/Input/Input.tsx (hunk 9 above)
Comment (joaooliveiradev) [reply src/components/Input/Input.tsx:16]: Fixed in this [commit](https://github.com/joaooliveiradev/conduit-frontend/pull/43/commits/17b6cad8efe25edb11aa873c9aeb288047ff229b)

File: src/components/NewArticle/NewArticle.tsx (hunk 11 above)
Comment (joaooliveiradev) [reply src/components/NewArticle/NewArticle.tsx]: What do you think about this [way](https://github.com/joaooliveiradev/conduit-frontend/pull/43/commits/8a9c2ead395250545dce964d12e8be95fe3d79a5)

File: src/components/NewArticle/NewArticle.tsx (hunk 13 above)
Comment (joaooliveiradev) [reply src/components/NewArticle/NewArticle.tsx]: Fixed in this [commit](https://github.com/joaooliveiradev/conduit-frontend/pull/43/commits/6479fa461658d52d71ce7613a126c392e06e6ed5).

File: src/components/NewArticle/NewArticle.tsx (hunk 15 above)
Comment (joaooliveiradev) [reply src/components/NewArticle/NewArticle.tsx:194]: What do you think about this [way](https://github.com/joaooliveiradev/conduit-frontend/pull/43/commits/fb5c90044e649c511e4adfe6a820f2406f1acc0a).

File: src/pages/editor/index.tsx (hunk 16 above)
Comment (joaooliveiradev) [reply src/pages/editor/index.tsx]: What do you think about this [way](https://github.com/joaooliveiradev/conduit-frontend/pull/43/commits/4f2fdca46910b916e569f906ddbdb9010e917c8e).

Comment (joaooliveiradev) [reply src/pages/editor/index.tsx]: I removed the Wrapper component in Editor page on this top commit, but i figured out that was important to have it, so i make another [commit](https://github.com/joaooliveiradev/conduit-frontend/pull/43/commits/1b6101acec87b701cda1184b4c4e8748dac303d8).

**Pull request merged**